---
title: "Lyme Disease and the the Tick Microbiome"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##### *Carter Merenstein*

## Background


In this project I looked at data from a study on the microbiome of ticks in the Eastern United States. The paper by Van Treuen et al. (2015) collected two species of ticks, *Ixodes scapularis* and *Ixodes affinis*, from 5 states, Connecticut, New York, Virginia, North Carolina, and South Carolina. *I. scapularis* is the primary vector for Lyme disease, which is reported in over 30,000 patients per year in the United States (Centers for Disease Control and Prevention). Lyme disease is caused by the bacteria *Borrelia burgdorferi*, and a human becomes infected after a bite from an infected tick (Steere, 2001).

In addition to *B. burgdorferi*, ticks can carry hundreds of other bacteria that make up their microbiome. Previous studies of the microbiomes of other animals, from humans down to insects, have shown that their microfauna can have great impacts on their health (Narasimhan and Fikrig, 2015; Turnbaugh et al., 2009). Importantly, in vectors of human disease, including ticks, certain members of the microbiome can prevent the disease agent from colonizing the host (Weiss and Aksoy, 2011). A famous example of this is an *Enterobacter sp.* in mosquitoes that can prevent the mosquito from carrying the malaria-causing protozoan (Cirimotich et al., 2011).

In this project, I will use data on the microbiomes of 92 ticks, first for an exploratory data analysis and subsequently to try to identify bacteria associated with lower *Borrelia* prevalence. The microbiome data was originally in the form of DNA sequences of the bacterial 16S rRNA gene, a standard biomarker for identifying bacteria. This was used to identify Operational Taxonomic Units (OTUs), which are clusters of bacteria that share 97% sequence identity, in the program QIIME. The R analysis in this project starts from a table with OTUs as variables and ticks as observations.



```{r pressure, echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=6}
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(tidyr))
suppressPackageStartupMessages(library(randomForest))
suppressPackageStartupMessages(library(stringr))


#this global is used for random forests later
# at 50000 the models get stable, but 500 is usefull for kntting fast.
NUM_TREES = 500

# This vector of colors is used for the stacked bar graphs.
c25 <- c("#6A3D9A", # purple
         "dodgerblue2","#E31A1C", # red
         "green4",
         "#FF7F00", # orange
         "Grey","gold1",
         "steelblue4","#FB9A99", # lt pink
         "palegreen2",
         "#CAB2D6", # lt purple
         "maroon","#FDBF6F", # lt orange
         "khaki2",
         "orchid1","deeppink1","blue1",
         "darkturquoise","green1","yellow4","yellow3","skyblue2",
         "darkorange4","brown", "gray70")

# These two other vectors are used for certain subseted stacked bar graphs
# They have colors removed to ensure that color-coding is consistent
c24 <- c25
c24[24] = "gray70" ## for females
cN <- c25[-13][-20] ## for negatives

# Data are in long format; each row is a tick, each column an OTU
ticks <- read.csv("data/ALL_Ticks.csv")
ticks_long <- ticks #this gets used later
ticks_long.orig <- ticks_long
ticks <- ticks %>% gather(OTU, Count, 2:ncol(ticks))
ticks.orig <- ticks
# Turn count into percent
stacks <- group_by(ticks, Tick, OTU) %>%  summarize(count = sum(Count))  %>% 
  filter(count > 0) %>% 
  inner_join(group_by(ticks, Tick) %>% summarize( Total = sum(Count))) %>% 
  mutate(Percent = ((count/Total) * 100) )

#taxonomy file has full taxonomic information of each OTU
taxonomy <- read.csv("data/OTU_Taxonomy.txt", sep = "\t") %>% tbl_df() %>% 
  mutate(Genus = gsub('[()0-9 ]', '', Genus)) %>% 
  mutate(Family = gsub('[()0-9 ]', '', Family)) %>% 
  mutate(Order = gsub('[()0-9 ]', '', Order))

# get OTU in same format
taxonomy <- mutate(taxonomy, OTU = gsub('[A-Z,a-z.]', '', OTU))
stacks <- mutate(stacks, OTU = gsub('[A-Z,a-z.]', '', OTU))

# the Taxon variable tells us the lowest identified taxonomical information
stacks <- inner_join(stacks, taxonomy, by="OTU") %>% 
  mutate(Taxon = ifelse(nchar(Genus) < 4, ifelse(nchar(Family) < 4, ifelse(nchar(Order) < 4, as.character(taxonomy), Order), Family), Genus)) %>% 
  mutate(OTU_taxon = paste(OTU, Taxon, sep = "; "))

# this is a totally janky section that is necessary because of some unidentified bug. Somewhere along the line the percents got messed up, this fizes them
stacks <- group_by(stacks, Tick) %>% summarize(Tot = sum(count)) %>% 
    inner_join(stacks, by = "Tick") %>% 
    mutate(p2 = count/Tot) %>% 
    mutate(Percent = p2 * 100)

# for stacked bar chart, we want only the top 24 most common, everything else gets put into "other"
most_common_df <- stacks %>% group_by(OTU_taxon) %>% 
    summarize(Total = sum(Percent)) %>%  top_n(24)
stacks_collapsed <- mutate(stacks, OTU_Collapsed = ifelse(OTU_taxon %in% most_common_df$OTU_taxon, OTU_taxon, "Other")) %>% 
  group_by(Tick, OTU_Collapsed)  %>% summarize(Percent = sum(Percent))

# the names tell what state they're from
# this was initially necessary because the metadata wasn't availible
# Now I could use the metadata file mf.txt, but this code works a well
NC_affinis = c("BLA", "HAL", "JON", "WAS")
NC_scapularis = c("FT", "Ixs")
SC = c("DRGF", "DRGM", "DRHF")

stacks_collapsed <- mutate(stacks_collapsed, region = gsub('[0-9.]+', '', Tick))

stacks_collapsed <- mutate(stacks_collapsed, region = ifelse(region %in% NC_affinis, "NC_a", region)) %>% 
  mutate(region = ifelse(region %in% NC_scapularis, "NC_s", region)) %>% 
  mutate(region = ifelse(region == "RJB", "VA", region)) %>% 
  mutate(region = ifelse(region %in% SC, "SC", region))

stacks_collapsed$region = as.factor(stacks_collapsed$region)

#Without further ado: stacked bar chart
stacked_bar_chart <- ggplot(stacks_collapsed, aes(x = Tick, y = Percent, fill = OTU_Collapsed)) +
    geom_bar(stat = "identity", color = "black") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    theme(legend.position="bottom", text=element_text(size=14), legend.key.size=unit(.25, "cm")) +
    scale_fill_manual(values = c25) +
      labs(fill = "OTU; Taxon") + 
    guides(fill = guide_legend(ncol = 3)) +
    theme(legend.background = element_rect(color = "Black", size = .5, linetype = "solid"))
```


## Exploratory Data Analysis

```{r , echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=6}

stacked_bar_chart

```
This stacked bar chart graphically represents the microbiome of each tick. The top 24 most common OTUs are shown and all others are grouped into the grey "other" category. Each vertical bar represents a single tick. OTUs are even along with the lowest identified taxonomy (o = order, f = family, g = genus).

The most noticeable trend is that some ticks have extremely low diversity, being made up of almost entirely one OTU. The two taxa that dominate are in the genus *Rickettsia*, and the family *Enterobacteriaceae*.

```{r , echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=4}

metadata <- read.csv("data/mf.txt", sep = '\t') %>% 
  rename(Tick = X.SampleID)
sex <- metadata %>%  select(`Tick`, `Sex`)

stacks_collapsed <- inner_join(stacks_collapsed, sex)

# it just looks better when they're two separate graphs, not faceted
f_collapsed = filter(stacks_collapsed, Sex == "F")
m_collapsed = filter(stacks_collapsed, Sex == "M")

stacked_bar_f <-ggplot(f_collapsed, aes(x = Tick, y = Percent, fill = OTU_Collapsed)) +
  geom_bar(stat = "identity", color = "black") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(legend.position="none" ,text=element_text(size=14)) +
  scale_fill_manual(values = c24) + 
  facet_grid(.~ region, scale="free", space="free") +
  labs(x = "Female Ticks")

stacked_bar_f
```

```{r , echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=6}
stacked_bar_m <-ggplot(m_collapsed, aes(x = Tick, y = Percent, fill = OTU_Collapsed)) +
  geom_bar(stat = "identity", color = "black") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(legend.position="bottom", text=element_text(size=14), legend.key.size=unit(.25, "cm")) +
  scale_fill_manual(values = c25) +
  facet_grid(.~ region, scale="free", space="free") +
  labs(x = "Male Ticks") +
  labs(fill = "OTU; Taxon") + 
    guides(fill = guide_legend(ncol = 3)) +
    theme(legend.background = element_rect(color = "Black", size = .5, linetype = "solid"))
stacked_bar_m

```


#### Differences in microbiome by Sex and Location

Some more differences become clear when we separate out male and female ticks, and facet by location collected. For example, we see that the females generally have a higher burden of Rickettsia (purple) than their male counterparts. It appears as though females are less diverse in general, which means that we might want to separate out males and females in future analysis. There are also pretty noticeable differences between state/species; *I. scapularis* ticks collected in North Carolina are dominated by an OTU in the family *Enterobacteriaceae* that is rare in other states. Again, this suggests that we will need to separate appropriately by species and state for future analysis.


```{r , echo=FALSE, message=FALSE, warning=FALSE, fig.width=10, fig.height=5}
# we really should subsampe so all have same total count, but the data I got were already subsampled, so the totals are pretty close to equal for all
ticks <- inner_join(ticks, sex, by="Tick")
species_richness <- group_by(ticks, Tick, Sex) %>%  summarize(Num_species = sum(Count > 0))
richness_histogram <- ggplot(data=species_richness, aes(x = Num_species)) +
    geom_histogram(bins=10) +
    facet_wrap(~Sex) +
    ggtitle("Species Richness by Sex")
richness_histogram
means <- species_richness %>% group_by(Sex) %>%
    summarize(mean = mean(Num_species))
se <- species_richness %>% group_by(Sex) %>%
    summarize(mean = sd(Num_species)/sqrt(n()))
```
A histogram of species richness (the number of OTU per tick) shows that the microbomes of male ticks are indeed more diverse than those of female ticks. Males had an average of 82 +/- 5.7 OTUs, females had 63 +/- 3.7 OTUs.

```{r location pca, echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=6}
# this part was done after I found the metadata file, so the region is done a bit differently than above
OTUs <- ticks_long[,2:ncol(ticks_long)]
just_ticks <- ticks_long[1]
tick.pca <- prcomp(OTUs, center=TRUE, scale.=TRUE)
#summary(tick.pca)
ticks_meta <- left_join(just_ticks, metadata, by = "Tick") #this seems redundant, but it puts it in the same order as ticks.pca, which is important
ticks_meta <- cbind(ticks_meta, tick.pca$x[1:nrow(ticks_meta), 1:2])

pca_location <- ggplot(ticks_meta, aes(x = PC1, y = PC2, fill = species_and_state), color = "black") + 
  geom_point(size = 4, pch = 21, alpha=0.5) + scale_fill_manual(values = c25) +
  ggtitle("PCA of all ticks by species and state")#+ geom_text(hjust = 0, nudge_x = 0.25)
#pca_location

```

###Principal Component Analysis

Another way we can compare microbiomes is with principal component analysis (PCA). This puts the microbiome on an x,y coordinate plane using axes that attempt to summarize as much of the variation as possible. The two best axes only were able to capture 16.7% of the variation in OTU abundance between ticks, meaning it only captures the biggest differences.

We see below that the male ticks spread apart much further than the females. This shows that there is a much greater diversity between male ticks (known as beta diversity), in addition to greater diversity within each male (alpha diversity).

Looking at the same PCA plot colored by species and state we see that *I. scapularis* ticks collected in Virginia show the greatest variation. In general ticks are fairly close to others collected in the same place. The most extreme case of this is *I. scapularis* from North Carolina, shown below as the red circles that are all on top of each other. These are probably so tightly clustered because they share the *Enterobacteriaceae* OTU (colored blue in the stacked bars) in high abundance.

Because the females cluster so strongly in the PCA of all ticks, a second set of PCA axes were made from only the female ticks. The first two axes in this case accounted for 94% of the variation between ticks, meaning they group ticks about as well as is possible. Coloring this PCA plot by species and state we see a similar pattern as we saw in the males; scapularis from NC cluster closest, and Virginia shows the greatest variation.

```{r , echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=6}
#initiallly this was combined with the one above, using shape, but it was just too much informaiton
pca_sex <- ggplot(ticks_meta, aes(x = PC1, y = PC2, fill = Sex), color = "black") + 
  geom_point(size = 4, pch = 21, alpha=0.5) +
  ggtitle("PCA of all ticks by sex")
pca_sex

```

```{r , echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=6}
pca_location

```

```{r , echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=6}
#let's try to see the differences better by running the pca on just females, since they were the ones clumping

female_ticks = ticks_meta %>%  filter(Sex == "F")
female_long <- ticks_long %>%  filter(Tick %in% female_ticks$Tick)
female_OTUs <- female_long[,2:ncol(female_long)]
f.pca <- prcomp(female_OTUs, center=T, scalue.=T)
#summary(f.pca)

female_ticks <- cbind(female_ticks, f.pca$x[1:nrow(female_ticks), 1:2])


f_pca_location <- ggplot(female_ticks, aes(x = PC1, y = PC2, fill = species_and_state), color = "black") + 
  geom_point(size = 4, pch = 21, alpha=0.5) + scale_fill_manual(values = c25) +
  ggtitle("PCA of female ticks by species and state") #+ geom_text(hjust = 0, nudge_x = 0.25)
f_pca_location

```

###Random Forest

The randomForest R package is a basic machine learning technique that builds a model for sorting data using repeated subsetting and cross validating. Here, we create a model to predict the state and species of a tick given the abundance of each OTU. The error rate is the percent at which the model misclassifies a tick, and the confusion matrix gives how often each type of tick was classified as each other type. As its name implies, a random forest model is non-deterministic, so the error rates and confusion matrix below may change with each knit, but with a large enough number of replicate trees the results will be fairly stable.

Here we see an error rate of around 24%. This is largely from *I. affinis* ticks in North Carolina and *I. scapularis* from New York. 81% of NC affinis ticks were misclassified as scapularis from Virginia. This tells us that the microbiomes of these two groups are highly similar almost indistinguishable. Ticks from NY were misclassified as either CT or VA 100% of the time.

Ticks from Virginia were successfully classified 100% of the time, despite having greater beta diversity as observed in the PCA. This is useful, since VA was the largest sample size. If the differences between locations may cloud out a signal in our comparison of *Borrelia* positives and negatives, we might be able to reduce noise by looking just within Virginia ticks.

```{r , echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=6}
sp_st <- ticks_meta %>%  select(Tick, species_and_state) 
df_ticks.rf <- inner_join(sp_st, ticks_long, by = "Tick") %>% 
  mutate(species_and_state = str_replace(species_and_state, "affinis", "a")) %>% 
  mutate(species_and_state = str_replace(species_and_state, "scapularis", "s")) %>% 
  mutate(species_and_state = as.factor(species_and_state))
#typeof(df_ticks.rf$species_and_state)

#at ntree 50,000 the fit seems fairly stable. Too far below this and it changes every time you run it.
fit <- randomForest(species_and_state ~ ., data=df_ticks.rf[,2:ncol(df_ticks.rf)], ntree=NUM_TREES)
fit
```
## Differences in ticks with and without *Borrelia*

The main objective of this project is to identify bacteria that may be preventing *Borrelia burgdorferi* (the Lyme disease bacteria) from colonizing a tick. I do this now by comparing the microbiomes of each and trying to find OTUs that are associated with a reduction in *Borrelia* prevalence. Below are two sets of stacked bar charts, negative ticks on top, and positives on bottom. In this view no major differences really stand out, except for the obvious that positives have the orange *Borrelia* OTU. Still, it's worth showing the stacked format again in case anything obvious did pop out.  

``` {r, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 12, fig.height=4}
# this Tick_borrelia_count.csv file is generated from the make_borrelia_file.R
# it was cleaner to just keep it in the separate R file and include the .csv

borrelia_counts <- read.csv("data/Tick_borrelia_count.csv") %>% 
  rename(borrelia_ct = Count)
borrelia_counts = borrelia_counts[-1]

stacks_collapsed <- inner_join(stacks_collapsed, borrelia_counts) %>% 
    mutate(borrelia = ifelse(borrelia_ct > 0, "positive", "negative"))

# single plot gets too wide, it's a lot easier to read when it's split up top and bottom than faceted left and right
# I know you can facet top bottom, but it just didn't look as good, so this is sorta the custom way.
# the color scales are manually coordinated to be the same, even though negatives are missing 2 OTUs
p_collapsed <- filter(stacks_collapsed, borrelia == "positive")
n_collapsed <- filter(stacks_collapsed, borrelia == "negative")

stacked_bar_n <- ggplot(n_collapsed, aes(x = Tick, y = Percent, fill = OTU_Collapsed)) +
  geom_bar(stat = "identity", color = "black") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    theme(legend.position="none", text=element_text(size=14)) +
    scale_fill_manual(values = cN) +
    labs(x = "Negative Ticks") +
    facet_grid(~region, space="free", scales= "free")
stacked_bar_n
```

``` {r, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 12, fig.height=6}
stacked_bar_p <- ggplot(p_collapsed, aes(x = Tick, y = Percent, fill = OTU_Collapsed)) +
  geom_bar(stat = "identity", color = "black") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    theme(legend.position="bottom", text=element_text(size=14), legend.key.size=unit(.25, "cm")) +
    scale_fill_manual(values = c25) +
      labs(fill = "OTU; Taxon", x = "Positive Ticks") + 
    guides(fill = guide_legend(ncol = 3)) +
    theme(legend.background = element_rect(color = "Black", size = .5, linetype = "solid")) +
      facet_grid(~region, space="free", scales= "free")

stacked_bar_p

```

###Random Forest to predict *Borrelia* abundance

In order to start identifying key differences in the microbiome of ticks with and without *Borrelia*, I again used a random forest model. This time, instead of classifying on a binary variable, I used the model to predict the abundance of OTUs in the genus Borrelia. I split by males and females because the above EDA showed considerable differences in males and females.

The model for the male ticks was able to explain 5-10% of the variance in *Borrelia* abundance. While this is low, it is still more useful than the female model, which had a negative % Var explained and therefore was worse than random.


``` {r, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 12, fig.height=6}
#some things here are rdundat

substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}

ticks_with_meta <- inner_join(stacks, metadata, by = "Tick") %>% 
  select(`Tick`, `Percent`, `OTU_taxon`, `species_and_state`, `state`, `Sex`)

#there's some punctuation in the names that randomForest doesn't like, but idk what it is so I get rid of all punctuation

df_borrelia <- ticks_with_meta %>% 
    mutate(OTU_taxon = gsub('[[:punct:] ]', '', paste("OTU", OTU_taxon, sep="")))  %>% 
  filter(substrRight(OTU_taxon, 8) !="Borrelia") 

df_borrelia <- inner_join(df_borrelia, borrelia_counts, by="Tick")
df_b_long <- df_borrelia %>%  spread(OTU_taxon, Percent, fill = 0)

male_ticks <- filter(df_b_long, Sex == "M")
female_ticks <- filter(df_b_long, Sex == "F")

# Yes, we should use percent of borrelia, not count for this. But, as mentioned earlier, these data are already subsampled (via. QIIME) such that all ticks have almost the same number of observations. It would be an easy fix but it's too late now.
male.rf <- randomForest(borrelia_ct ~ ., data=male_ticks[,5:ncol(male_ticks)], ntree = NUM_TREES)
male.rf

female.rf <- randomForest(borrelia_ct ~ ., data=female_ticks[,5:ncol(female_ticks)], ntree = NUM_TREES)
female.rf
#varImpPlot(female.rf)
```
#### OTUs predictive of *Borrelia* abundance

One useful feature of a randomForest approach is that our model can tell us what variables (OTUs) were most influential as predictors using a variable importance plot. In the case of a regression model, this plot can show us the variables most strongly correlated with the outcome variable.

Here we see that the 4 most important OTUs are all closely related, in the family *Oxalobacteraceae* (the genus *Jathinobacterium* is in this family). This suggests that these correlations might be the result of an actual biological interaction rather than random noise, as similar species are showing a similar pattern.

Unfortunately for us, these species are all positively correlated with *Borrelia* abundance. This means that even if it represents a biologically relevant interaction, these *Oxalobacteraceae* are actually facilitating *Borrelia* colonization, rather than preventing it.  


``` {r, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 12, fig.height=6}

varImpPlot(male.rf)

ggplot(male_ticks, aes(x = OTU305668gJanthinobacterium, y = borrelia_ct)) + geom_point() + stat_smooth(method = "lm") +
  labs(y = "Borrelia abundance", x = "OTU305668; g__Janthinobacterium (%)") 

```


####Classificaiton model

To get a better sense of how the model is working, I collapsed the Borrelia abundance into just positives and negatives. This lets us use random forest to classify ticks as positive or negative, and shows us a more helpful breakdown of the error we're getting.

Unfortunately, we see that our error rates for both males and females are worse than if we were just guessing based off of percent positives. This suggests that there's just a lot of noise in the data that's unavoidable and will throw off our model. One problem is also that there is a wide range in the *Borrelia* burden of positive ticks, from <1% to >25% of the whole microbiome being *Borrelia*. In this binary model we treat these ticks the same, but this difference will directly affect the remaining microbiome and introduce high variation within the positive group.

To try to reduce noise I also tried a model of just the males from Virginia, hoping to eliminate the effect of location. However, this model had an even worse error rate than the all male model.  

All Males:
``` {r, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 12, fig.height=6}


ticks_long_yn <- mutate(df_b_long, borrelia_ct = ifelse(borrelia_ct > 0, "positive", "negative")) %>% 
    mutate(borrelia_ct = as.factor(borrelia_ct)) %>% 
    rename(Borrelia = borrelia_ct)
rf_data <- ticks_long_yn[,-1][,-2][,-2]

male_ticks_yn <- filter(ticks_long_yn, Sex == "M")
female_ticks_yn <- filter(ticks_long_yn, Sex == "F")

male.rf_yn <- randomForest(Borrelia ~ ., data=male_ticks_yn[,5:ncol(male_ticks_yn)], ntree = NUM_TREES)
male.rf_yn
#varImpPlot(male.rf_yn)
```
All Females:
``` {r, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 12, fig.height=6}

female.rf_yn <- randomForest(Borrelia ~ ., data=female_ticks_yn[,5:ncol(female_ticks_yn)], ntree = NUM_TREES)
female.rf_yn
#varImpPlot(female.rf_yn)

```
Virginia Males:
``` {r, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 12, fig.height=6}

# the state and species didn't improve it so it's here but commented out

rf_data_m <- male_ticks_yn[,-1][,-2][,-2] #this includes state_and_species

# Does including state in the formula help?
male.rf_yn <- randomForest(Borrelia ~ ., data=rf_data, ntree = NUM_TREES)
#male.rf_yn
#varImpPlot(male.rf_yn)

VA_males <- filter(male_ticks_yn, species_and_state == "scapularis_va")
va.male.rf.yn <- randomForest(Borrelia ~ ., data=VA_males[,5:ncol(male_ticks_yn)], ntree = NUM_TREES)
va.male.rf.yn 
#varImpPlot(male.rf_yn)

```


### Difference in Frequency

Because the random forest models were not very successful in identifying bacteria that are negatively associated with *Borrelia* presence, I tried a second approach. Here I compare the frequency at which each OTU is found in ticks with and without *Borrelia*. This is an easy way to find OTUs that commonly co-occur with *Borrelia*, and those that are rarely found occupying the same tick as *Borrelia*. If a type of bacteria were able to prevent *Borrelia* from colonizing then it would be found more often in negative ticks and rarely in positive ticks.

Like the models above, I separated by sex in order to reduce the noise from sex differences. I also tested just the Virginia males to reduce location bias.

``` {r, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 12, fig.height=6}
otus_present <- inner_join(ticks_with_meta, borrelia_counts, by="Tick")
otus_present <- otus_present %>% 
  mutate(borrelia_ct = ifelse(borrelia_ct>0, "positive", "negative")) %>% 
  rename(Borrelia = borrelia_ct) %>% 
  filter(substrRight(OTU_taxon, 8) !="Borrelia") %>% 
  spread(OTU_taxon, Percent, fill = 0) #spread and gather re-intorudces 0s

#splitting it up made it work, idk why.
otus_present <- gather(otus_present, OTU_taxon, Percent, 6:ncol(otus_present)) %>% 
  mutate(present = ifelse(Percent > 0, 1, 0))


f_ticks <- otus_present %>%  filter(Sex == "F")
m_ticks <- otus_present %>%  filter(Sex == "M")

f_ticks <- f_ticks %>% group_by(Borrelia, OTU_taxon) %>%
    summarize(frequency = mean(present))

f_ticks <- spread(f_ticks, Borrelia, frequency) %>% 
  mutate(diff = positive - negative)

ggplot(f_ticks %>%  top_n(20, abs(diff)), aes(x = reorder(OTU_taxon, diff), y = diff, fill = as.factor(ifelse(diff > 0, "Positive", "Negative")))) +
  geom_bar(stat = "identity") +
  coord_flip() + guides(fill=FALSE) +
  labs(y = "<- less common with Borrelia, more common with Borrelia ->", x = "OTU; Taxonomy") +
  theme(text = element_text(size=14)) +
  ggtitle("All Females")

m_ticks <- m_ticks %>% group_by(Borrelia, OTU_taxon) %>%
  summarize(frequency = mean(present))

m_ticks <- spread(m_ticks, Borrelia, frequency) %>% 
  mutate(diff = positive - negative)

ggplot(m_ticks %>%  top_n(20, abs(diff)), aes(x = reorder(OTU_taxon, diff), y = diff, fill = as.factor(ifelse(diff > 0, "Positive", "Negative")))) +
  geom_bar(stat = "identity") +
  coord_flip() + guides(fill=FALSE) +
  labs(y = "<- less common with Borrelia, more common with Borrelia ->", x = "OTU; Taxonomy")+
theme(text = element_text(size=14)) +
  ggtitle("All Males")

VA_males_diff <- otus_present %>%  filter(Sex == "M") %>% 
  filter(species_and_state == "scapularis_va") %>% 
  group_by(Borrelia, OTU_taxon) %>%
  summarize(frequency = mean(present)) %>% 
  spread( Borrelia, frequency) %>% 
  mutate(diff = positive - negative)

ggplot(VA_males_diff %>%  top_n(20, abs(diff)), aes(x = reorder(OTU_taxon, diff), y = diff, fill = as.factor(ifelse(diff > 0, "Positive", "Negative")))) +
  geom_bar(stat = "identity") +
  coord_flip() + guides(fill=FALSE) +
  labs(y = "<- less common with Borrelia, more common with Borrelia ->", x = "OTU; Taxonomy") +
theme(text = element_text(size=14)) +
  ggtitle("Virginia Males")

VA_males_chi <- otus_present %>% filter(Sex == "M") %>% 
  filter(species_and_state == "scapularis_va")

num_VAm <- VA_males_chi %>% group_by(Borrelia, Tick) %>% summarize(junk = n()) %>% 
    group_by(Borrelia) %>% summarize(num = n())
pos = as.numeric(num_VAm[2,2])
neg = as.numeric(num_VAm[1,2])

VA_males_chi <- VA_males_chi %>% 
  group_by(Borrelia, OTU_taxon) %>%
  summarize(frequency = mean(present)) %>% 
  spread( Borrelia, frequency) %>% 
  mutate(negative = negative * neg) %>% 
  mutate(positive = positive * pos)

```

#### Significance in differences

The OTU 4480861 (g__*Catenibacterium*) was found more often in ticks without *Borrelia* than those with. A Chi Squared test found that this difference is actually significant to a 95% cutoff (p = 0.035, below). However, in these data we had over 300 OTUs that I test here for differences in prevalence. With a alpha of 0.05 we would still expect to have more than 15 OTUs with a p<0.05, even if the data were totally random. To get around this, we need to adjust our p value to reflect the number OTUs tested. This adjustment is called the Bonferoni correction and simply involves multiplying the p value (0.035) by the number of variables tested. Following this correction we no longer have any OTUs that are found significantly more often in the absence of *Borrelia*.


``` {r, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 12, fig.height=6}

#from the graph we get the most differentially present
most_sig <- VA_males_chi %>%  filter(OTU_taxon == "4480861; g__Catenibacterium") 
#most_sig
#way cleaner to just bind these manually
chi_df <- rbind (c(3, 0), c(neg-3, pos-0))
chisq.test(chi_df)

```

## Conclusions


In this project I demonstrated that female ticks are less diverse than males, both in terms of species richness and in beta diversity. This was primarily driven by an increase in the abundance of two OTUs of *Rickettsia* and *Enterobacteraceae* in most females. Principal component analysis showed that ticks from some locations cluster closely together (North Carolina), while others show high diversity between ticks (Virginia). A random forest algorithm demonstrated that *I. affinis* ticks from North Carolina are similar to *I. scapularis* in Virginia, though ticks from Virginia were always successfully distinguished as such.

These findings in the exploratory phase were used to drive a comparison of ticks with and without *Borrelia*. The most meaningful trend that I found was a positive correlation between abundance of bacteria in the family *Oxalobacteraceae*. As discussed previously, when testing a large number of variables for possible correlations, false positives are likely. However, finding 4 false positives that all happen to be in the same biological family is considerably less likely. This suggests to me that there is some biological reason why *Borrelia* and *Oxalobacteraceae* abundances are related. This doesn't necessarily suggest that these bacteria directly interact, however. It is possible that they are vulnerable to similar environmental stressors or thrive in similar growth conditions.

Further attempts to identify bacteria that are negatively associated with *Borrelia* were unsuccessful. These data are very noisy, in that there are many types of bacteria that almost certainly have no relation to *Borrelia*. For example, some bacteria live only in the ovaries of the tick, while *Borrelia* colonizes the midgut. The more unrelated bacteria included in these data the more any signal gets drowned out by the noise. A more thorough analysis could filter out species that are known not to occupy a similar niche as *Borrelia*, though a large majority of bacteria identified here have not been previously studied in ticks. One way around this in future studies is to dissect the tick prior to DNA sequencing in order to specifically study the microbiome of the midgut.

#### References
Centers for Disease Control and Prevention Lyme Disease.
Cirimotich, C.M., Dong, Y., Clayton, A.M., Sandiford, S.L., Souza-Neto, J.A., Mulenga, M., and Dimopoulos, G. (2011). Natural Microbe-Mediated Refractoriness to Plasmodium Infection in Anopheles gambiae. Science 332, 855–858.
Narasimhan, S., and Fikrig, E. (2015). Tick microbiome: the force within. Trends Parasitol. 31, 315–323.
Steere, A.C. (2001). Lyme disease. N. Engl. J. Med. 345, 115–125.
Treuren, W.V., Ponnusamy, L., Brinkerhoff, R.J., Gonzalez, A., Parobek, C.M., Juliano, J.J., Andreadis, T.G., Falco, R.C., Ziegler, L.B., Hathaway, N., et al. (2015). Variation in the Microbiota of Ixodes ticks with geography, species and sex. Appl. Environ. Microbiol. AEM.01562–15.
Turnbaugh, P.J., Hamady, M., Yatsunenko, T., Cantarel, B.L., Duncan, A., Ley, R.E., Sogin, M.L., Jones, W.J., Roe, B.A., Affourtit, J.P., et al. (2009). A core gut microbiome in obese and lean twins. Nature 457, 480–484.
Weiss, B., and Aksoy, S. (2011). Microbiome influences on insect host vector competence. Trends Parasitol. 27, 514–522.




